homeassistant/components/sensor/ambient_station.py

```diff
@@ -0,0 +1,205 @@
+        sensor_params = SENSOR_TYPES[condition]
+        name = sensor_params[SENSOR_NAME]
+        units = sensor_params[SENSOR_UNITS]
+        if isinstance(units, list):
+            units = sensor_params[SENSOR_UNITS][UNIT_SYSTEM[sys_units]]
+
+        sensor_list.append(AmbientWeatherSensor(station_data, condition, name,
+                                                units))
+
+    add_devices(sensor_list)
+
+
+class AmbientWeatherSensor(Entity):
+    """Representation of a Sensor."""
+
+    def __init__(self, station_data, condition, name, units):
+        """Initialize the sensor."""
+        self._state = None
+        self.station_data = station_data
+        self._condition = condition
+        self._name = name
+        self._units = units
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def unit_of_measurement(self):
+        """Return the unit of measurement."""
+        return self._units
+
+    def update(self):
+        """Fetch new state data for the sensor.
+
+        This is the only method that should fetch new data for Home Assistant.
+        """
+        _LOGGER.debug("Getting data for sensor: %s", self._name)
+        data = self.station_data.get_data()
+        if data is None:
+            # update likely got throttled and returned None, so use the cached
+            # data from the station_data object
+            self._state = self.station_data.data[self._condition]
+        else:
+            if self._condition in data:
+                self._state = data[self._condition]
+            else:
+                _LOGGER.warning("%s sensor data not available from the "
+                                "station", self._condition)
+
+        _LOGGER.debug("Sensor: %s | Data: %s", self._name, self._state)
+
+
+class AmbientStationData:
+    """Class to interface with ambient-api library."""
+
+    def __init__(self, api_key, app_key, interval):
+        """Initialize station data object."""
+        self._api_key = api_key
+        self._app_key = app_key
+
+        self._interval = interval
+
+        self._api_keys = {
+            'AMBIENT_ENDPOINT':
+            'https://api.ambientweather.net/v1',
+            'AMBIENT_API_KEY': self._api_key,
+            'AMBIENT_APPLICATION_KEY': self._app_key,
+            'log_level': 'DEBUG'
+        }
+
+        self.data = None
+        self._station = None
+        self._api = None
+        self._devices = None
+
+        self.get_data = Throttle(self._interval)(self.update)
+
+    def update(self):
+        """Get new data."""
+        # refresh API connection since servers turn over nightly
+        _LOGGER.debug("Getting new data from server")
+        new_data = None
+        self._connect_api()
+        time.sleep(2)   # need minimum 2 seconds between API calls
```

Comment:
@MartinHjelmare - The sleep is inside of the update because the _connect_api() counts as an API call as well as the station.get_data() method.  The server only allows requests every 2 seconds.

I had initially created the AmbientStationData object a single time, however the server turns over the API service once a day at an unknown time and you have to reconnect.  Therefore, the connection will go stale within 24 hours and there is no way to know this.  The best solution I could come up with is to re-instantiate the AmbientStationData object on every update.

Any ideas on how to overcome this issue?

Reply:
Then we need to convert `update` methods both here and on the entity to use asyncio.
```py
async def async_update(self):
    ...
    data = await self.station_data.get_data()

...

async def async_update(self):
    ...
    await self.hass.async_add_executor_job(self._connect_api)
    await asyncio.sleep(2)
    if self._station is not None:
        data = await self.hass.async_add_executor_job(self._station.get_data)
    ...
```